feat: Add Holidays feature to block availability on public holidays

dhairyashiil opened the pull request:
## What does this PR do?

Adds a new "Holidays" feature that allows users to automatically block their availability on public holidays based on their country.


Fixes #8918
Fixes CAL-6865

### Features
- Country-based holidays: Support for 36 countries via Google Calendar public holiday calendars
- Per-holiday toggles: Users can enable/disable specific holidays
- Booking conflict detection: Warns users if they have existing bookings on holidays they're enabling
- OOO integration: Shows warning when creating OOO entries that overlap with holidays

### Implementation
- Google Calendar API integration: Fetches holidays from Google's public holiday calendars (no OAuth required, uses API key)
- Database caching: HolidayCache Prisma model stores fetched holidays with configurable TTL (default 7 days via HOLIDAY_CACHE_DAYS env)
- New UserHolidaySettings Prisma model for user preferences (country selection, disabled holidays)
- TRPC endpoints for settings management and conflict checking
- Holidays tab in OOO page
- Integration with getUserAvailability for blocking time slots

### Environment Variables:
- GOOGLE_CALENDAR_API_KEY(Required) - Google Calendar API key for fetching public holiday calendars
- HOLIDAY_CACHE_DAYS(Not Required) - Cache duration in days (default: 7)



## Screenshots/Videos
you can also see _booking conflict warning_ 




https://github.com/user-attachments/assets/0eda2d84-ce27-462d-b759-e5b53109afce







_ooo warning:_

<img width="2555" height="1364" alt="Screenshot 2025-12-03 at 8 09 25 PM" src="https://github.com/user-attachments/assets/64ba17fa-f731-4085-891a-75549ab3c4a2" />


_Updated holiday emoji_

light:

<img width="1361" height="852" alt="Screenshot 2025-12-03 at 10 15 20 PM" src="https://github.com/user-attachments/assets/c98d02c5-8c04-474e-97b3-7a9b80cda86a" />


Dark:


<img width="1478" height="862" alt="Screenshot 2025-12-03 at 10 14 58 PM" src="https://github.com/user-attachments/assets/3d1b3930-21c2-4390-9062-be296409bd22" />


Booking page:


<img width="1437" height="770" alt="Screenshot 2025-12-03 at 10 21 51 PM" src="https://github.com/user-attachments/assets/e3834102-b862-4cc4-bf61-c8d4463a067c" />



### Test coverage
- Unit tests for `HolidayService`


### Supported Countries
US, UK, Canada, Australia, Germany, France, Spain, Netherlands, Italy, Brazil, Mexico, India, Ireland, New Zealand, Sweden, Norway, Denmark, Belgium, Austria, Switzerland, Japan, China, South Korea, Singapore, Hong Kong, Taiwan, Thailand, Malaysia, Indonesia, Philippines, Vietnam, Russia, Poland, Greece, Portugal, Finland, South Africa, Iran

refer: https://gist.githubusercontent.com/mattn/1438183/raw/612d3178bef9afa832adc1ff82062bb50aba152c/google-calendar-list.txt
update: used this: https://gist.github.com/dhoeric/76bd1c15168ee0ee61ad3bf1730dcb65

dhairyashiil committed:
feat: add holidays feature for automatic availability blocking- Add UserHolidaySettings model for storing user preferences- Generate static holiday data for 20 countries using date-holidays- Create HolidayService for runtime holiday queries- Add TRPC router with endpoints for country selection and holiday toggles- Create Holidays tab UI in Availability page with conflict warnings- Integrate holiday blocking into getUserAvailability calculation- Show holiday indicator on blocked dates in booker page- Add warning in OOO modal when dates overlap with holidays

dhairyashiil committed:
feat: add optimizations, tests, and code quality improvements

- Add memoization/caching to HolidayService for better performance
- Optimize checkConflicts DB query with OR conditions for specific dates
- Add HolidayService unit tests (10 tests)
- Add error handling with Alert component for failed queries
- Memoize HolidayListItem component to prevent unnecessary re-renders
- Extract magic numbers into constants.ts
- Use TRPCError consistently in handlers
- Add missing i18n keys for error messages
- Update handlers to follow Cal.com patterns (default exports, minimal comments)
- Add regeneration instructions in constants

dhairyashiil committed:
refactor: replace static JSON with Google Calendar API integration

- Add GoogleHolidayService to fetch holidays from Google Calendar public calendars
- Add HolidayCache Prisma model for caching API responses
- Add GOOGLE_CALENDAR_API_KEY and HOLIDAY_CACHE_DAYS env variables
- Support 38 countries via Google Calendar holiday calendars
- Update HolidayService methods to async with database caching
- Update all TRPC handlers for async holiday methods
- Fix UI to display holiday dates correctly
- Remove static holidays.json and generate script

dhairyashiil committed:
use we instead of calcom in i18n message

dhairyashiil committed:
address cubics comments

CarinaWolli reviewed (dismissed):
We want to move this to `/my-account/out-of-office` instead and it shows too many holidays that are not real public holidays

dhairyashiil committed:
move holidays from availability to ooo

dhairyashiil committed:
public holidays filter for holidays

dhairyashiil committed:
follow i18n _one and _other pattern

dhairyashiil committed:
remove holiday feature flag

dhairyashiil committed:
revert lint command code change

dhairyashiil committed:
revert lint command code change 2.0

dhairyashiil committed:
revert lint command code change 3.0

dhairyashiil committed:
bye bye my christmas emoji :crying-emoji

dhairyashiil committed:
remove comments

keithwillcode reviewed (dismissed):
The only tests seem to be for the HolidayService. We'll need new tests to cover its usage in places like getUserAvailablity.ts since there's logic around what's returned by this.

ferki commented:
Thanks for your effort on this!

> ### Supported Countries
> US, UK, Canada, Australia, Germany, France, Spain, Netherlands, Italy, Brazil, Mexico, India, Ireland, New Zealand, Sweden, Norway, Denmark, Belgium, Austria, Switzerland, Japan, China, South Korea, Singapore, Hong Kong, Taiwan, Thailand, Malaysia, Indonesia, Philippines, Vietnam, Russia, Poland, Greece, Portugal, Finland, South Africa, Iran
> 
> refer: https://gist.githubusercontent.com/mattn/1438183/raw/612d3178bef9afa832adc1ff82062bb50aba152c/google-calendar-list.txt

I noticed this list does not include the calendar for Hungarian public holidays, which I know exists among Google calendars.

Apparently the cited source feels outdated, it even mentions “Created 14 years ago” on its main gist link: https://gist.github.com/mattn/1438183

Please consider using a current list of Google public holiday calendars instead.

I found this collection which states “Last active 2 days ago”, listing 237 calendars, and also including instructions how to get a fresh list: https://gist.github.com/dhoeric/76bd1c15168ee0ee61ad3bf1730dcb65

Hope this helps!

eunjae-lee commented on packages/lib/holidays/GoogleHolidayService.ts:
```diff
@@ -0,0 +1,216 @@
+    return data.items
+      .filter((event) => {
+        const desc = event.description?.toLowerCase() || "";
+        return desc.includes("public holiday");
```
I'm asking with no knowledge or clue at all, but is it guaranteed that they include "public holiday" text in the description? (for non-English speaking countries?)

eunjae-lee commented on packages/lib/holidays/GoogleHolidayService.ts:
```diff
@@ -0,0 +1,216 @@
+      await this.refreshCache(countryCode, year);
+    }
+
+    const cached = await prisma.holidayCache.findMany({
```
I'd create a HolidayRepository and use it here instead of raw prisma code

eunjae-lee commented on packages/lib/holidays/GoogleHolidayService.ts:
```diff
@@ -0,0 +1,216 @@
+  }
+}
+
+export const GoogleHolidayService = new GoogleHolidayServiceClass();
```
why not just exporting the class and let the caller instantiate it? You can refer to other examples of services, for example, `packages/lib/server/service/ApiKeyService.ts`

eunjae-lee commented on packages/trpc/server/routers/viewer/holidays/toggleHoliday.handler.ts:
```diff
@@ -0,0 +1,68 @@
+
+export async function toggleHolidayHandler({ ctx, input }: ToggleHolidayOptions) {
+  const userId = ctx.user.id;
+  const { holidayId, enabled } = input;
```
Q: is this `holidayId` an id of HolidayCache?

eunjae-lee reviewed (commented):
Left some feedback and questions

dhairyashiil committed:
refactor(holidays): add repository pattern, split services, and add tests

- Create HolidayRepository for database operations
- Split GoogleHolidayService into GoogleCalendarClient and HolidayCacheService
- Add dependency injection to HolidayService and HolidayCacheService
- Update TRPC handlers to use HolidayRepository
- Add tests for HolidayRepository and calculateHolidayBlockedDates
- Update calendar IDs to use official holiday format (244 countries + religions)

dhairyashiil committed:
chore: merge upstream/main

dhairyashiil committed:
fix: address PR review feedback

- Remove unused date-holidays package
- Add pluralization for and_more_holidays_with_conflicts translation
- DRY: spread GOOGLE_RELIGIOUS_HOLIDAY_CALENDARS into GOOGLE_HOLIDAY_CALENDARS
- Add select to userHolidaySettings query in getUserAvailability
- Optimize checkConflicts with pre-computed timestamps

dhairyashiil committed:
Merge branch 'main' into feat/holidays-tab

eunjae-lee commented on turbo.json:
```diff
@@ -76,8 +76,10 @@
+    "GOOGLE_CALENDAR_API_KEY",
     "GOOGLE_LOGIN_ENABLED",
     "GOOGLE_WEBHOOK_TOKEN",
+    "HOLIDAY_CACHE_DAYS",
```
not sure if we need to add this as an env var. what about just a const variable somewhere like constants.ts ?

eunjae-lee commented on packages/trpc/server/routers/viewer/holidays/checkConflicts.handler.ts:
```diff
@@ -0,0 +1,119 @@
+    return { conflicts: [] };
+  }
+
+  const startDate = new Date();
```
let's avoid having logics and prisma calls directly within the handler. instead, move them into services & repositories.

eunjae-lee commented on packages/lib/holidays/constants.ts:
we may not need a functionality to control this with env var. a const value might be enough until we get a demand.

dhairyashiil committed:
refactor(holidays): apply proxy pattern and move logic to service

- Rename HolidayCacheService to HolidayServiceCachingProxy (proxy pattern)
- Remove HOLIDAY_CACHE_DAYS env var, use constant directly
- Add isSupportedCountry() method to HolidayService
- Add getUserSettings() and updateSettings() to HolidayService
- Move toggleHoliday logic from handler to service
- Move checkConflicts logic from handler to service
- Add findBookingsInDateRanges() to HolidayRepository
- Simplify all handlers to just call service methods

eunjae-lee commented on packages/trpc/server/routers/viewer/holidays/toggleHoliday.handler.ts:
```diff
@@ -0,0 +1,35 @@
+    return await holidayService.toggleHoliday(ctx.user.id, input.holidayId, input.enabled);
+  } catch (error) {
+    throw new TRPCError({
+      code: "BAD_REQUEST",
```
It's not about the wrong input to this handler, so we may use `INTERNAL_SERVER_ERROR` instead.

dhairyashiil committed:
feat(bookings): add backend validation to prevent booking on holidays

Adds explicit holiday conflict validation during booking creation to
handle the race condition where a host enables a holiday after a guest
selects a date but before they submit the booking.

Changes:
- Add checkHolidayConflict validation with HolidayRepository integration
- Integrate ensureNoHolidayConflict in RegularBookingService
- Add BookingOnHoliday error code with proper HTTP 400 response
- Handle holiday error display in BookEventForm with name interpolation
- Hide trace ID for expected validation errors
- Add unit tests for holiday conflict validation

dhairyashiil committed:
resolve merge conflicts

dhairyashiil commented:
Added check to prevent booking on holidays:


https://github.com/user-attachments/assets/5f770b43-d755-422c-a078-347b7e61f56a

dhairyashiil committed:
fix failing type check

dhairyashiil committed:
fix: address PR review comments for holiday feature

- Change error code from BAD_REQUEST to INTERNAL_SERVER_ERROR in
  toggleHoliday handler (errors are internal failures, not bad input)
- Refactor ensureNoHolidayConflict to use Promise.all for parallel
  user checking instead of sequential loop

dhairyashiil committed:
refactor: use Promise.all with logging in loop for holiday check

- Check all users in parallel using Promise.all
- Log conflicts inside the loop as they are detected
- Wait for all checks to complete before throwing error

dhairyashiil committed:
fix(holidays): use host timezone for holiday conflict checks

The holiday feature was checking booking dates against holidays using
server/local timezone instead of the host's timezone. This caused
bookings near midnight boundaries to incorrectly pass or fail the
holiday check.

Example: A booking at Dec 24th 8PM UTC (which is Dec 25th in IST)
was not being blocked for an Indian host with Christmas as a holiday.

Changes:
- Add .utc() to holiday date formatting for consistency
- Fetch host's timezone in checkHolidayConflict
- Convert booking time to host's timezone before comparison
- Add findUserSettingsWithTimezone to HolidayRepository
- Update error message to clarify it's the host's local time
- Add timezone edge case tests

dhairyashiil committed:
fix(holidays): align holiday blocking with OOO pattern for consistent timezone handling

- Simplify calculateHolidayBlockedDates to match OOO pattern using dayjs.utc()
- Fix date range query to use full day bounds (startOfDay/endOfDay) so holidays
  stored at midnight UTC are correctly found during booking validation
- Remove separate checkHolidayConflict booking-time validation - holidays now
  block through oooExcludedDateRanges like OOO does
- Remove getHolidayOnDate from HolidayService (no longer needed)
- Remove findUserSettingsWithTimezone from HolidayRepository (no longer needed)
- Clean up related tests

Holiday blocking now works exactly like OOO:
1. Holidays are added to datesOutOfOffice in calculateHolidayBlockedDates
2. buildDateRanges processes them via processOOO with .tz(timeZone, true)
3. oooExcludedDateRanges excludes those dates from availability
4. ensureAvailableUsers uses oooExcludedDateRanges to block bookings

This ensures consistent timezone handling where Dec 25th blocks all hours
of Dec 25th in the host's timezone, regardless of booker's timezone.

dhairyashiil committed:
update test

dhairyashiil committed:
chore: merge upstream/main into feat/holidays-tab

dhairyashiil committed:
update .env.example file

eunjae-lee reviewed (approved):
looks good and it's working for me

hbjORbj reviewed (approved):
Please address my comments in a follow-up PR, it's important feedback. Approving because @CarinaWolli needs this merged now

emrysal committed:
Merge branch 'main' into feat/holidays-tab
